lti_tool/views.py

```diff
@@ -0,0 +1,309 @@
+            profile.save()
+
+            self.profile = profile
+        login(self.request, self.user)
+        self.request.session["lti-launch-id"] = message_launch.get_launch_id()
+
+        if message_launch.is_resource_launch():
+            launch_params = self.message_launch_data["https://purl.imsglobal.org/spec/lti/claim/custom"]
+            instance = CourseInstance.objects.get(
+                course__url=launch_params["course_slug"],
+                url=launch_params["instance_slug"]
+            )
+            roles = self.message_launch_data["https://purl.imsglobal.org/spec/lti/claim/roles"]
+            is_course_teacher = Enrollment.objects.filter(
+                    course_instance=instance,
+                    user_profile=UserProfile.objects.get(user=self.user),
+                    role=Enrollment.ENROLLMENT_ROLE.TEACHER
+                ).exists()
+            if "http://purl.imsglobal.org/vocab/lis/v2/membership#Learner" in roles and not is_course_teacher:
+                enrolment_success = instance.enroll_student(self.user)
+            if "exercise_path" in launch_params:
+                url_params = {
+                    k: launch_params[k] for k in ["exercise_path", "module_slug", "course_slug", "instance_slug"]
+                }
+                return redirect("lti-exercise", **url_params)
+            elif "module_slug" in launch_params:
+                url_params = {
+                    k: launch_params[k] for k in ["module_slug", "course_slug", "instance_slug"]
+                }
+                return redirect("lti-module", **url_params)
+            elif "course_slug" in launch_params and "instance_slug" in launch_params:
+                url_params = {
+                    k: launch_params[k] for k in ["course_slug", "instance_slug"]
+                }
+                return redirect("lti-course", **url_params)
+            else:
+                pass # invalid params
+        elif message_launch.is_deep_link_launch():
+            return redirect("lti-select-content")
+        else:
+            pass # invalid launch type - should not reach
+
+class LtiSessionMixin(BaseMixin):
+
+    @xframe_options_exempt
+    def dispatch(self, *args, **kwargs):
+        return super().dispatch(*args, **kwargs)
+
+    def parse_lti_session_params(self):
+        launch_id = self.request.session.get("lti-launch-id", None)
+        tool_conf = get_tool_conf()
+        if launch_id != None:
+            self.message_launch = DjangoMessageLaunch.from_cache(
+                launch_id,
+                self.request,
+                tool_conf,
+                launch_data_storage=get_launch_data_storage()
+            )
+            self.message_launch_data = self.message_launch.get_launch_data()
+
+class LtiInstanceView(InstanceView, LtiSessionMixin):
+    
+    access_mode = ACCESS.ENROLLED
```

Comment:
There should be two empty lines between each of the different classes in this file.